src/webapp/github.py

```diff
@@ -0,0 +1,120 @@
+import base64
+import json
+import re
+import urllib.error
+import urllib.request
+
+api_baseurl = "https://api.github.com"
+
+# review actions
+APPROVE         = 'APPROVE'
+REQUEST_CHANGES = 'REQUEST_CHANGES'
+COMMENT         = 'COMMENT'
+
+
+def mk_github_authstr(user, passwd):
+    raw = '%s:%s' % (user, passwd)
+    return base64.encodebytes(raw.encode()).decode().replace('\n', '')
+
+def api_path2url(api_path, **kw):
+    fmtstr = re.sub(r':([a-z]+)\b', r'{\1}', api_path)
+    path = fmtstr.format(**kw)
+    url = api_baseurl + path
+    return url
+
+class GitHubAuth:
+    api_user = None
+    api_token = None
+    api_authstr = None
+    logger = None
+
+    def __init__(self, api_user, api_token, logger=None):
+        self.api_user = api_user
+        self.api_token = api_token
+        self.api_authstr = mk_github_authstr(api_user, api_token)
+        self.logger = logger
+
+    def elog(self, msg):
+        if self.logger:
+            self.logger.error(msg)
+
+    def dlog(self, msg):
+        if self.logger:
+            self.logger.debug(msg)
+
+    def github_api_call(self, method, url, data):
+        if data is not None:
+            data = json.dumps(data).encode()
+        req = urllib.request.Request(url, data)
```

Comment:
Does doing `req = urllib.request.Request(url, data, method=method)` not work? Then you wouldn't have to use a lambda below.

Reply:
How about that.

In python2, `urllib2.Request` (which python3 moved to `urllib.request.Request`), the `__init__` method did not take a `method` parameter, and it was somewhat annoyingly required that `req.get_method` be set to a function that would return the method to be used.  Thus the lambda.

I didn't realize they added a method parameter.  Apparently this was added in Python 3.3.

So, while this `github_api_call` goes about things in a way that is also python2 compatible, yeah it looks cleaner to just pass in a `method` parameter to `Request`, since we only care about python3.6+.